### ISSUE TITLE
Config issues on gunpowder dataloading branch

### ISSUE THREAD

**Issue by Christianfoley** [mehta-lab/microDL#197]
@Soorya19Pradeep brought up some issues with certain config parameters that did not function as intended for training/preprocessing:

1. 'channel_ids' on preprocessing flags. Specifying a single integer or a list of integers produces some interpreting error
2. Simply changing architecture and depth to match a 2D config produces an assertion error, depth != 5. 

Both of these need to be fixed to allow 2D training.

**Comment by Christianfoley** [comment]
Config used to produce this error can be found at `/hpc/projects/CompMicro/projects/infected_cell_imaging/VirtualStaining/config_files/torch_gpLoading_2023_01/`

**Comment by Soorya19Pradeep** [comment]
Hello @Christianfoley. Is it possible this error is caused by my zarr file lacking channel names? I understand the io-util.py under utils folder enumerates the channel names which is used down the line in creating and storing mask channels during preprocessing. I haven't been naming my channels on storing after the XZY registration of channels.

**Comment by Soorya19Pradeep** [comment]
I have made a small change in preprocessing file that fixed the issue.
commit tag: 045de3bb35f07a0445b77d310b152b3ba084603e

**Comment by Soorya19Pradeep** [comment]
I have been encountering another issue with the preprocessing related to the number of positions found for preprocessing. There are 31 FOVs, 5 channels and 9 z-slices in the current dataset I am working on located here:
`/hpc/projects/CompMicro/projects/infected_cell_imaging/VirtualStaining/VirtualStain_NuclMem_A549_2023_02_07/Input_Nucl/A549_20X.zarr`
The preprocessing find 54 positions instead. The code doesn't run till end as it cannot find all the positions it accounted for initially to proceed with the segmentation step.
![image](https://user-images.githubusercontent.com/101817974/218594033-e8f47638-7cea-4ad6-8d01-a1f172161d2e.png)

**Comment by Christianfoley** [comment]
As you mentioned if the position metadata does not match the data, the preprocessing will not be able to track all of the positions/arrays and will terminate poorly when it can't find an array it's looking for. 

Currently, the zarr reader that the preprocessing uses is embedded in the `microDL` code and is outdated. Once Ziwen releases a "candidate" for IoHub, we should be able to implement some zarr-store validation to make sure you we have to sit through 60% of preprocessing just to find the metadata was corrupted.

**Comment by Soorya19Pradeep** [comment]
@Christianfoley , this is a valid point. On discussing with @mattersoflight we were thinking of fixing the original metadata, but this will not be a fix for situations where we eliminate some positions from our original dataset or doesn't complete set imaging for multiple time points. I too think the solution you mentioned is the way to go. Should we start a new issue for this?

**Comment by Christianfoley** [comment]
Since this is more of an initial formatting issue than part of the actual staining pipeline, I think this should probably be accomplished in a script, rather than as part of the core codebase. @mattersoflight what do you think?

**Comment by Soorya19Pradeep** [comment]
@Christianfoley the issue with the zarr metadata does not exist anymore. I found it was a fault from my image registration script. I was able to correct the image registration script and correct the zarr metadata. I have been able to run preprocessing on the dataset successfully. Thank you for your patience!

I did encounter a different aspect though. In my case I want to train the model to predict the nuclear stain from the phase image. The original dataset contains 5 channels in this order. 
["Retardance", "Orientation", "Phase", "DAPI", "Y5"]
Here DAPI is the nuclear stain and Y5 is the membrane stain. In my config file I have specified input channel as [2] (phase) and target channel as [3] (nuclear stain). If I do not mention channel_ids: [3] in preprocessing part of the config, instead have channel_ids: -1, it masks the retardance (channel 0) and stores it in the zarr. Should we make sure to mention the channels in preprocessing to avoid this error and maybe document this?

**Comment by mattersoflight** [comment]
@Soorya19Pradeep , @Christianfoley good to see these tests and debugging of the gunpowder data loader. 

**selecting positions and recording them accurately in metadata:**
As a principle, the metadata should be fixed at the source. In other words,
* If the metadata recorded with acquisition is erroneous, it should be edited in raw data. 
* If the metadata generated by a reconstruction script is erroneous, it should be corrected with a script.

I think you are on the right track. 

**Channel names in config**
The data loader should allow users to specify inputs and outputs by channel names. Let's not use channel index, because they are an easy source of confusion. As the data is preprocessed sometimes the channel indices can shift too. `iohub` is now public and has features to add new channels. It will soon have features to delete and rewrite channels. 

The above should be a separate issue. There will be other issues that we discover during testing. 

In this issue, I suggest you make the config work by hand and document the bugs and feature requests.

**Comment by Soorya19Pradeep** [comment]
Updates on 2D model training:
- In preprocessing, during flatfielding mean of image computation gets negative values in DAPI image and therefore as a check skips the flatfielding of the channel. An empty array is returned resulting in code terminating with error.
- In training, the voxel size used to check alignment of data by gunpowder dataloader is read to be (1,1), but the window size expected is (1,256,256) as specified in config.

**Comment by Soorya19Pradeep** [comment]
Path to copy of dataset which was used for 2D model testing, which caused the above errors:
`/hpc/projects/CompMicro/projects/infected_cell_imaging/VirtualStaining/VirtualStain_NuclMem_A549Cell_2022_12_09/Input_nucl/A549_20X.zarr`
Path to the original dataset:
`/hpc/projects/CompMicro/projects/infected_cell_imaging/Image_preprocessing/Exp_2022_12_09_A549NuclMemStain/A549_20X_2D/Registered/A549_20X.zarr`
